scripts/skills/perks/perk_rf_whirling_death.nut

```diff
@@ -1,46 +1,168 @@
 		}
+	}
+
+	function onTurnStart()
+	{
+		this.m.Stacks = 0;
+	}
 
-		return true;
+	function onCombatFinished()
+	{
+		this.skill.onCombatFinished();
+		this.m.Stacks = 0;
 	}
-	
-	function onAnySkillUsed( _skill, _targetEntity, _properties )
+
+	function onPayForItemAction( _skill, _items )
 	{
-		if (this.isEnabled() && (this.m.IsForceEnabled || (_skill.getItem() != null && _skill.getItem().getID() == this.getContainer().getActor().getMainhandItem().getID())))
+		foreach (item in _items)
 		{
-			_properties.DamageDirectAdd += 0.01 * (_targetEntity == null ? this.m.MaxBonus : ::Math.rand(this.m.MinBonus, this.m.MaxBonus));
+			if (_item != null && _item.getSlotType() == ::Const.ItemSlot.Mainhand)
+			{
+				this.m.Stacks = 0;
+				return;
+			}
 		}
 	}
-});
 
+	function doExtraAttack( _skill, _primaryTargetTile )
+	{
+		local user = this.getContainer().getActor();
+		local myTile = user.getTile();
+
+		local targets = [];
+		for (local i = 0; i < 6; i++)
+		{
+			if (!myTile.hasNextTile(i))
+				continue;
+
+			local nextTile = myTile.getNextTile(i);
+			if (!nextTile.IsOccupiedByActor || nextTile.isSameTileAs(_primaryTargetTile))
+				continue;
+
+			local entity = nextTile.getEntity();
+			if (!entity.isAlliedWith(user) && entity.getCurrentProperties().getReach() < user.getCurrentProperties().getReach() && _skill.onVerifyTarget(myTile, nextTile))
+			{
+				targets.push(entity)
+			}
+		}
+
+		if (targets.len() == 0)
+			return;
+
+		local target = ::MSU.Array.rand(targets);
+		local targetTile = target.getTile();
+
+		if (!user.isHiddenToPlayer() || targetTile.IsVisibleForPlayer)
+		{
+			this.getContainer().setBusy(true);
+			::Time.scheduleEvent(::TimeUnit.Virtual, 100, function ( _perk )
+			{
+				if (target.isAlive())
+				{
+					this.logDebug("[" + user.getName() + "] is attacking [" + target.getName() + "] with skill [" + _skill.getName() + "] due to " + _perk.m.Name);
+					if (!user.isHiddenToPlayer() && targetTile.IsVisibleForPlayer)
+					{
+						::Tactical.EventLog.log(::Const.UI.getColorizedEntityName(user) + " has " + _perk.m.Name);
+					}
+
+					this.m.IsPerformingExtraAttack = true;
+					_skill.useForFree(targetTile);
+					this.m.IsPerformingExtraAttack = false;
+				}
+				this.getContainer().setBusy(false);
+
+			}.bindenv(this), this);
+		}
+		else
+		{
+			if (target.isAlive())
+			{
+				this.logDebug("[" + user.getName() + "] is attacking [" + target.getName() + "] with skill [" + _skill.getName() + "] due to " + this.m.Name);
+
+				this.m.IsPerformingExtraAttack = true;
+				_skill.useForFree(targetTile);
+				this.m.IsPerformingExtraAttack = false;
+			}
+		}
+	}
+
+	function isSkillValid( _skill )
+	{
+		if (!_skill.isAttack() || _skill.isRanged())
+			return false;
+
+		if (this.m.RequiredWeaponType == null)
+			return true;
+
+		local weapon = _skill.getItem();
+		return !::MSU.isNull(weapon) && weapon.isItemType(::Const.Items.ItemType.Weapon) && weapon.isWeaponType(::Const.Items.WeaponType.Flail);
```

Comment:
You should use `this.m.RequiredWeaponType` variable here instead of hard coding Flail